mesh/mesh.hpp

```diff
@@ -381,6 +381,7 @@ class Mesh
    int FindCoarseElement(int i);
 
    /// Update the nodes of a curved mesh after refinement
+   /** If the Nodes GridFunction is defined, invokes NodesUpdated(). */
    void UpdateNodes();
```

Comment:
Hmm, `UpdateNodes` and `NodesUpdated` is a little confusing.

I guess `UpdateNodes` rebuilds the nodal FE space (which is required after refinement), whereas `NodesUpdated` just resets the geometric factors. Could `UpdateNodes` be given a clearer name (since it's not a public member function)? Or maybe just a little more explanation in the Doxygen comment.

Another question: should `NodesUpdated` call `SetVerticesFromNodes`?

Reply:
> Could `UpdateNodes` be given a clearer name (since it's not a public member function)? Or maybe just a little more explanation in the Doxygen comment.

Do you have specific suggestions? I'm not sure what more to say in the doxygen comment.

> Another question: should `NodesUpdated` call `SetVerticesFromNodes`?

It's a matter of convention, I suppose. Currently, `UpdateNodes` calls `SetVerticesFromNodes` before calling `NodesUpdated`. In other places, it depends -- e.g. if the user does not need/use the vertices, then calling `SetVerticesFromNodes` will be redundant. If they change the nodes and need to change the vertices too, they can call `SetVerticesFromNodes` directly to express that need.

Reply:
> Do you have specific suggestions? I'm not sure what more to say in the doxygen comment.

One potential name is `UpdateNodesAfterRefinement` or something similar.

In terms of the comment, I meant spelling out the difference explicitly, e.g. (assuming that I correctly understand the difference myself): `UpdateNodes` is called after topological changes to the mesh (e.g. refinement), whereas `NodesUpdated` should be called by the user after changing the mesh geometry/curvature by modifying the internal `Nodes` grid function.

(Or some variation thereof). I just think that with such similar function names, we should very explicitly disambiguate.

> It's a matter of convention, I suppose. Currently, UpdateNodes calls SetVerticesFromNodes before calling NodesUpdated. In other places, it depends -- e.g. if the user does not need/use the vertices, then calling SetVerticesFromNodes will be redundant. If they change the nodes and need to change the vertices too, they can call SetVerticesFromNodes directly to express that need.

Sounds reasonable to me.

Reply:
I updated the doxygen docs to point out the main difference between `UpdateNodes` and `NodesUpdated` in the docs for both methods. Let me know if the updates look ok to you.

I did not change the name of `UpdateNodes` -- I'm sure it will be that helpful and we'll probably want to keep the old name for some time for backward compatibility. However, if you insist, I can change it.

Reply:
> I updated the doxygen docs to point out the main difference between UpdateNodes and NodesUpdated in the docs for both methods. Let me know if the updates look ok to you.

Looks good, thanks.

> I did not change the name of UpdateNodes -- I'm sure it will be that helpful and we'll probably want to keep the old name for some time for backward compatibility. However, if you insist, I can change it.

No, I don't insist on it, what we have now seems fine. I was hoping that because `UpdateNodes` is not public, we could change the name without a deprecation period for backwards compatibility (I suppose because it is protected, it could technically be breaking, but I suspect nobody external calls this function).